src/TestContext.elm

```diff
@@ -197,47 +212,76 @@ createWithNavigationAndJsonStringFlags flagsDecoder onRouteChange program initia
                         }
 
 
-update : msg -> TestContext msg model -> TestContext msg model
+update : msg -> TestContext msg model effect -> TestContext msg model effect
 update msg (TestContext result) =
     TestContext <|
         case result of
             Err err ->
                 Err err
 
-            Ok ( program, model ) ->
+            Ok ( program, ( model, _ ) ) ->
                 Ok
                     ( program
                     , program.update msg model
-                        |> Tuple.first
                     )
 
 
-clickButton : String -> TestContext msg model -> TestContext msg model
-clickButton buttonText (TestContext result) =
+simulateHelper : String -> (Query.Single msg -> Query.Single msg) -> ( String, Json.Encode.Value ) -> TestContext msg model effect -> TestContext msg model effect
+simulateHelper functionDescription findTarget event (TestContext result) =
     case result of
         Err err ->
             TestContext <| Err err
 
-        Ok ( program, model ) ->
+        Ok ( program, ( model, _ ) ) ->
+            let
+                targetQuery =
+                    program.view model
+                        |> Query.fromHtml
+                        |> findTarget
+            in
+            -- First check the target so we can give a better error message if it doesn't exist
             case
-                program.view model
-                    |> Query.fromHtml
-                    |> Query.find
-                        [ Selector.tag "button"
-                        , Selector.containing [ Selector.text buttonText ]
-                        ]
-                    |> Test.Html.Event.simulate Test.Html.Event.click
-                    |> Test.Html.Event.toResult
+                targetQuery
+                    |> Query.has []
+                    |> Test.Runner.getFailureReason
             of
-                Err message ->
+                Just reason ->
                     TestContext <|
-                        Err (SimulateFailed ("clickButton " ++ toString buttonText) message)
+                        Err (SimulateFailedToFindTarget functionDescription reason.description)
+
+                Nothing ->
+                    -- Try to simulate the event, now that we know the target exists
+                    case
+                        targetQuery
+                            |> Test.Html.Event.simulate event
+                            |> Test.Html.Event.toResult
+                    of
+                        Err message ->
+                            TestContext <|
+                                Err (SimulateFailed functionDescription message)
+
+                        Ok msg ->
+                            update msg (TestContext result)
 
-                Ok msg ->
-                    update msg (TestContext result)
 
+simulate : (Query.Single msg -> Query.Single msg) -> ( String, Json.Encode.Value ) -> TestContext msg model effect -> TestContext msg model effect
```

Comment:
Might be good to add an example in the doc comment for this - I was initially confused about what `(String, Json.Encode.Value)` was supposed to represent until I realized that you can get these from `Test.Html.Event`